### PR TITLE
Add Hosts caching for canProvision()

### PR DESCRIPTION
**Pull request by scoheb**
[FIXES JENKINS-38815]

### CODE REVIEW

**Comment by scoheb** [comment]
@pjanouse

File: src/main/java/com/redhat/foreman/ForemanSharedNodeWorker.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+                    foremanCloud.updateHostData();
+                    LOGGER.debug("[COMPLETED] Updating data for ForemanSharedNodeCloud " + foremanCloud.getCloudName());
+                }
+            }
+        }
+    }
+
+    @Override
+    public long getRecurrencePeriod() {
+        return MIN;
```
Comment (pjanouse) [inline src/main/java/com/redhat/foreman/ForemanSharedNodeWorker.java]: We can make configurable this later, but isn't 1m period so aggressive?

File: src/main/java/com/redhat/foreman/ForemanSharedNodeCloud.java [hunk 2]
```diff
@@ -90,6 +90,7 @@
 
     private transient ForemanAPI api = null;
     private transient ForemanComputerLauncherFactory launcherFactory = null;
+    private transient Map<String, String> hostsMap = new HashMap<String, String>();
```
Comment (pjanouse) [inline src/main/java/com/redhat/foreman/ForemanSharedNodeCloud.java]: Isn't more save and robust use a `java.util.concurrent.atomic.AtomicReference<V>` here?

File: src/main/java/com/redhat/foreman/ForemanSharedNodeCloud.java [hunk 3]
```diff
@@ -300,14 +295,6 @@ private ForemanSharedNode provision(Label label) throws Exception {
...
     private String getHostToReserve(Label label) {
-        Map<String, String> hostsMap = null;
-        try {
-            hostsMap = getForemanAPI().getCompatibleHosts();
-        } catch (Exception e) {
-            LOGGER.error("Unhandled exception in getHostToReserve: ", e);
-            e.printStackTrace();
-            return null;
-        }
         Set<Map.Entry<String, String>> hosts = hostsMap.entrySet();
```
Comment (pjanouse) [inline src/main/java/com/redhat/foreman/ForemanSharedNodeCloud.java]: Possible NPE here is hostsMap is null (similar like is handled on L157).

File: src/main/java/com/redhat/foreman/ForemanSharedNodeCloud.java [hunk 4]
```diff
@@ -426,6 +413,15 @@ public Integer getSshConnectionTimeOut() {
         return sshConnectionTimeOut;
     }
 
+    public void updateHostData() {
+        try {
+            hostsMap = getForemanAPI().getCompatibleHosts();
+        } catch (Exception e) {
+            LOGGER.error(e);
```
Comment (pjanouse) [inline src/main/java/com/redhat/foreman/ForemanSharedNodeCloud.java]: We shouldn't use a stale data, what about to reset `hostsMap` to `null` in this case?

File: src/main/java/com/redhat/foreman/ForemanSharedNodeCloud.java [hunk 5]
```diff
@@ -426,6 +413,15 @@ public Integer getSshConnectionTimeOut() {
         return sshConnectionTimeOut;
     }
 
+    public void updateHostData() {
```
Comment (scoheb) [inline src/main/java/com/redhat/foreman/ForemanSharedNodeCloud.java]: Why? it is a void

**Comment by scoheb** [comment]
@pjanouse

**Comment by pjanouse** [comment]
Your changes are great, I'm for them.

Probably one minor change... What about to initialize `hostsMap` reference value as an empty `Map` right in declaration line? The same approach on L445 too - I mean never to set `hostMap` to `null` in the code (including if an `Exception` occurs), but to empty `Map`. In that case we are safe to remove checks on line 161 and 307, aren't we?

**Comment by scoheb** [comment]
I agree with ensuring we return an empty map. I thought you had preferred that we return a null :)

a new change on its way...

**Comment by scoheb** [comment]
@pjanouse

**Comment by pjanouse** [comment]
This revision is ready to go, I don't have any objection.

Anyway you are right, my original statement was different and thank you for patience:).

**Comment by pjanouse** [comment]
Hm.. probably one proposal - it seems to use `AsyncPeriodicWork` instead of `PeriodicWork` would be a better choice in our case, plus please add this constructor:
``` 
        public ForemanSharedNodeWorker() {
            super("ForemanSharedNodeWorker.Updater");
        }
```

And please remove lines L459...L461 from `getHostsMapData()`, this operation can be potentially a quite time spent and we don't want to block a caller. The real issue can be only during Jenkins start-up, but well, we can claim for awhile that we aren't able to provision asked slaves until the structure isn't constructed (which can take a quite long time in the production again and we don't want to block anybody by this).

**Comment by pjanouse** [comment]
Note: PR#20 affect this.

**Comment by pjanouse** [comment]
I did all necessary changes, including rebasing and incorporating PR#20.

Unfortunately I don't know how to update your original PR#17, so created PR#21...

This PR should be closed without merge.

Note: If there is a better way how to make a work-flow more smoothly in this case, please give me an advice.

**Comment by scoheb** [comment]
The PR was closed when I merged PR20...so what you did worked perfectly!

**Comment by pjanouse** [comment]
I guess you meant PR#21... that's great, I've never did this before, so good to know for the next time:).

**Comment by scoheb** [comment]
Yes. #21 . Shoot me an email as to why you chose java.util.logging as opposed to log4j. Curious to hear your thoughts

**Pull request merged**